Framework/DataHandling/src/MultiPeriodLoadMuonStrategy.cpp

```diff
@@ -0,0 +1,138 @@
+
+  auto sampleInformation =
+      LoadMuonNexusV2Helper::loadSampleInformationFromNexus(m_entry);
+  std::string mainFieldDirection =
+      LoadMuonNexusV2Helper::loadMainFieldDirectionFromNexus(m_entry);
+  double firstGoodData =
+      LoadMuonNexusV2Helper::loadFirstGoodDataFromNexus(m_entry);
+
+  for (int i = 0; i < m_workspaceGroup->getNumberOfEntries(); ++i) {
+    auto workspace =
+        std::dynamic_pointer_cast<Workspace2D>(m_workspaceGroup->getItem(i));
+    auto &run = workspace->mutableRun();
+    run.addProperty("main_field_direction", mainFieldDirection);
+    run.addProperty("FirstGoodData", firstGoodData);
+    run.addProperty("sample_temp", sampleInformation.temperature);
+    run.addProperty("sample_magn_field", sampleInformation.magneticField);
+  }
+}
+/**
+ * Loads the good frames data into each of the stored workspace objects
+ */
+void MultiPeriodLoadMuonStrategy::loadGoodFrames() {
+  NXInt goodframes = LoadMuonNexusV2Helper::loadGoodFramesDataFromNexus(
+      m_entry, MULTIPERIODSLOADED);
+  for (int i = 0; i < m_workspaceGroup->getNumberOfEntries(); ++i) {
+    auto workspace =
+        std::dynamic_pointer_cast<Workspace2D>(m_workspaceGroup->getItem(i));
+    auto &run = workspace->mutableRun();
+    run.removeProperty(GOODFRAMESPROP);
+    run.addProperty(GOODFRAMESPROP, goodframes[i]);
+  }
+}
+
+/**
+ * Loads detector grouping.
+ * If no entry in NeXus file for grouping, load it from the IDF.
+ * stored
+ * @returns :: TableWorkspace group containing each of the grouping tables
+ */
+Workspace_sptr MultiPeriodLoadMuonStrategy::loadDetectorGrouping() const {
+  // Each period could in theory have its own grouping, which is reflected in
+  // the nexus file which has (periods*numDetectors) entries in the Nexus
+  // grouping entry.
+  WorkspaceGroup_sptr tableGroup = std::make_shared<WorkspaceGroup>();
+  for (int i = 0; i < m_workspaceGroup->getNumberOfEntries(); ++i) {
+    int periodNumber = i + 1;
+    auto grouping = LoadMuonNexusV2Helper::loadDetectorGroupingFromNexus(
+        m_entry, m_detectors, MULTIPERIODSLOADED, periodNumber);
+    TableWorkspace_sptr table =
+        createDetectorGroupingTable(m_detectors, grouping);
+    // if any of the tables are empty we'll load grouping from the IDF
+    if (table->rowCount() == 0) {
+      m_logger.notice("Loading grouping information from IDF");
+      return LoadMuonNexusV2Helper::loadDefaultDetectorGrouping(
+          std::dynamic_pointer_cast<Workspace2D>(m_workspaceGroup->getItem(i)));
+    } else {
+      tableGroup->addWorkspace(table);
+    }
+  }
+  return tableGroup;
+}
+/**
+ * Performs time-zero correction on the loaded workspace.
+ */
+void MultiPeriodLoadMuonStrategy::applyTimeZeroCorrection() {
+  double timeZero = LoadMuonNexusV2Helper::loadTimeZeroFromNexusFile(m_entry);
+  for (int i = 0; i < m_workspaceGroup->getNumberOfEntries(); ++i) {
+    auto workspace =
+        std::dynamic_pointer_cast<Workspace2D>(m_workspaceGroup->getItem(i));
```

Comment:
After seeing this snippet repeated loads of times, could we just have a single free function somewhere called 

```
Workspace2d* castWsFromGroupIndex(WorkspaceGroup_sptr &wsGroup, size_t i){
    ws = std::dynamic_pointer_cast<Workspace2D>(wsGroup0>getItem(i));
    assert(ws); // Or throw
    return ws;
}
```

Then change all code sites (including the other files) to use this as it's much easier to read
`auto ws = castWsFromGroupIndex(wsGroup, i)`
Plus any future changes are done in one place then